Share the exception local ID table

methodmissing opened the pull request:
Introduces a new `ISEQ_SHARED_EXC_LOCAL_TBL` flag that facilitates conditional free of the locals table in `rb_iseq_free`. This allows for sharing a single constant ID table via `iseq_set_exception_local_table` for the lifetime of the process.

I don't like the fact that the allocation isn't freed, however in `gc.c` `id_to_obj_tbl` and `obj_to_id_tbl` for example aren't explicitly freed on exit either.

Debug counters dumped on exit from booting up a redmine server process - the exception local table was shared about `1500` times:

```
[RUBY_DEBUG_COUNTER]	lvar_get                      	       2121321
[RUBY_DEBUG_COUNTER]	lvar_get_dynamic              	        196824
[RUBY_DEBUG_COUNTER]	lvar_set                      	        240395
[RUBY_DEBUG_COUNTER]	lvar_set_dynamic              	         19189
[RUBY_DEBUG_COUNTER]	lvar_set_slowpath             	           866
[RUBY_DEBUG_COUNTER]	iseq_shared_exc_local_tbl     	          1595 <<<<<<<<<<<<<<
[RUBY_DEBUG_COUNTER]	gc_count                      	            54
[RUBY_DEBUG_COUNTER]	gc_minor_newobj               	            43
[RUBY_DEBUG_COUNTER]	gc_minor_malloc               	             0
[RUBY_DEBUG_COUNTER]	gc_minor_method               	             0
[RUBY_DEBUG_COUNTER]	gc_minor_capi                 	             0
[RUBY_DEBUG_COUNTER]	gc_minor_stress               	             0
[RUBY_DEBUG_COUNTER]	gc_major_nofree               	            11
```

nobu commented:
Good catch.

ko1 commented:
I need to use git blame who remained a TODO comment...

```
163d436d428 (ko1           2015-09-12 18:47:43 +0000  1364)     /* TODO: every id table is same -> share it.
163d436d428 (ko1           2015-09-12 18:47:43 +0000  1365)      * Current problem is iseq_free().
163d436d428 (ko1           2015-09-12 18:47:43 +0000  1366)      */
```

nobu commented:
I think the table can be const `ID` array.
And is the debug counter needed?

methodmissing commented:
Yes, the comment was left by @ko1 - stumbled on it while going through the local table implementation and decided to try to make it work. const `ID` array is fine and the debug counter can be :boom:

I can change it in a while

methodmissing commented:
Removed the debug counter and qualified the ID table data as constant, but did not push further for constant pointer too.

methodmissing commented:
Rebased trunk to handle merge conflict on `idERROR_INFO` from https://github.com/ruby/ruby/commit/54b93ef1ac4c81730508740ae213f6f52b0bec94

nobu commented:
Why is `shared_exc_local_tbl` still a pointer while it is never freed and its size is fixed?

methodmissing commented:
Forward fixed

nobu commented:
Comparison with the array address is sufficient instead of the new flag, isn't?
A predicate function may be better, I guess.

methodmissing commented:
Thanks for the continuous constructive feedback. Removed the flag and implemented comparison by address, however the const table now crosses compile units with it being defined in `compile.c` and the guard on local table free in `iseq.c`. I see `ruby_api_version` used in the same way in `iseq.h`, thus should be fine.

The patch is *much* smaller now though :smile:

nobu commented:
`RUBY_EXTERN` is to export against extension libraries, but this array is not; just `extern`.
As global symbols without `RUBY_EXTERN` may be visible or invisible from extension libraries depending on platforms, it needs a prefix, `ruby_`, `rb_`, `rb_iseq_`, and so on for example.

methodmissing commented:
Thanks for the great explanation and the `ẁhy`s. Pushed - preferred the `rb_iseq_` prefix.
